File: reverseinteger.js
//Objective is to reverse an integer. If the integer is negative, reverse
//and make it negative.

let num = -321


//O(n) solution where n is the number of digits in the number.

let temp = Math.abs(num)
let value = 0

//Similar to 'Happy Number'
while (temp > 0) {
    value = value * 10 + temp % 10
    temp = Math.floor(temp / 10)
}

if (value > Math.pow(2, 31) || value < Math.pow(-2, 31)) {
    return 0
} else if (num < 0) {
    return -value
} else {
    return value
}

File: reverseinteger.java
//Java Solution

class Solution {
    public int reverse(int x) {
        int result = 0;
        int prevResult = 0;
        
        while (x != 0) {
            result = (result * 10) + (x % 10);
            
            if((result - x % 10) / 10 != prevResult){
                return 0;
            }
            
            prevResult = result;
            x = x / 10;
        }
        
        return result;
    }
}